onnxruntime/core/framework/session_state.cc

```diff
@@ -450,26 +451,27 @@ Status SessionState::GeneratePatternGroupCache(const std::vector<std::reference_
       }
 
       ORT_ENFORCE(exe_plan->allocation_plan[ml_value_idx].alloc_kind == AllocKind::kAllocate);
-      ORT_ENFORCE(exe_plan->allocation_plan[ml_value_idx].program_counter_start.size() == exe_plan->allocation_plan[ml_value_idx].program_counter_end.size());
 
-      for (size_t index = 0; index < exe_plan->allocation_plan[ml_value_idx].program_counter_start.size(); index += 1)
-        ORT_ENFORCE(exe_plan->allocation_plan[ml_value_idx].program_counter_start[index] <= exe_plan->allocation_plan[ml_value_idx].program_counter_end[index]);
-
-      mem_planner.TraceAllocation(ml_value_idx, exe_plan->allocation_plan[ml_value_idx].program_counter_start,
-                                  exe_plan->allocation_plan[ml_value_idx].program_counter_end, size);
+      const auto& counter = exe_plan->allocation_plan[ml_value_idx].program_counter;
+      for (size_t index = 0, end = counter.Starts().size(); index < end; index += 1) {
+        mem_planner.TraceAllocation(ml_value_idx, counter, size);
+      }
```

Comment:
Why is *ml_value_idx* being traced in a loop? It should just be traced once. This makes me wonder if we should put an assert in GeneratePatterns to make sure no ml-value is traced more than once otherwise that will increase the footprint of static allocation buffer for no good reason. Your change will pass CI builds but it will definitely cause regression on performance side of things for which we do not have a good gate in place ... there is some powerBI dashboard but no sure how helpful that will be.

Reply:
Great catch! Will fix.